From 731dd3e4c279e70f6fb6c8a3167ddec91210220b Mon Sep 17 00:00:00 2001
Message-Id: <731dd3e4c279e70f6fb6c8a3167ddec91210220b.1433790354.git.chang-joon.lee@intel.com>
From: Tan Wei <weix.z.tan@intel.com>
Date: Tue, 2 Jun 2015 09:21:10 +0800
Subject: [PATCH] SQUASHME! [VPG]: drm/i915: correct lock sequence in
 drm_mode_obj_set_property_ioctl

The lock sequence in drm_mode_obj_set_property_ioctl will cause deadlock
with drm_modeset_lock_all().

Fixed the issue by using the same lock sequence in drm_modeset_lock_all:
mode_config's mutex -> connection_mutex -> crtc's mutex

This patch should be squashed into:
    commit e79bcf7a2a9342f471a49a2a6403f65ff0227a9e
    Change-Id: Ife6cba9fc6338966d6b8924544b52437389946c3
    Author: Akashdeep Sharma <akashdeep.sharma@intel.com>
    Date:   Wed May 20 00:27:40 2015 +0530

        REVERTME [VPG]: drm: Take single CRTC lock

Issue: IMINAN-37777
Change-Id: Ie8d8556213da78c771355a4ae0a264524d9dc6ab
Signed-off-by: Tan Wei <weix.z.tan@intel.com>
---
 drivers/gpu/drm/drm_crtc.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/drm_crtc.c b/drivers/gpu/drm/drm_crtc.c
index c522643..68d521f 100644
--- a/drivers/gpu/drm/drm_crtc.c
+++ b/drivers/gpu/drm/drm_crtc.c
@@ -4200,11 +4200,10 @@ int drm_mode_obj_set_property_ioctl(struct drm_device *dev, void *data,
 			DRM_ERROR("CRTC from connector,CRTC={id=%d props=%d}\n",
 				crtc->base.id, crtc->base.properties ?
 					crtc->base.properties->count : 0);
-			drm_modeset_lock(&crtc->mutex, NULL);
 			mutex_lock(&config->mutex);
+			drm_modeset_lock(&crtc->mutex, NULL);
 			ret = drm_mode_connector_set_obj_prop(arg_obj, property,
 							arg->value);
-			mutex_unlock(&config->mutex);
 		} else {
 			DRM_ERROR("No crtc from connector, lock all\n");
 			drm_modeset_lock_all(dev);
@@ -4227,8 +4226,11 @@ int drm_mode_obj_set_property_ioctl(struct drm_device *dev, void *data,
 		break;
 	}
 
-	if (crtc)
+	if (crtc) {
 		drm_modeset_unlock(&crtc->mutex);
+		if (arg_obj->type == DRM_MODE_OBJECT_CONNECTOR)
+			mutex_unlock(&config->mutex);
+	}
 	else
 		drm_modeset_unlock_all(dev);
 out:
-- 
1.9.1

